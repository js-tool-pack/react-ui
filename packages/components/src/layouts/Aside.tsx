import React from 'react';
import { BaseLayoutsProps } from './layout.types';
import { getComponentClass } from '@pkg/shared';
import { getClassNames } from '@tool-pack/basic';

const rootClass = getComponentClass('aside');

export const Aside: React.FC<BaseLayoutsProps> = (props) => {
  const { children, className, ...rest } = props;
  return (
    <aside
      {...rest}
      className={getClassNames(rootClass, {
        [className as string]: className,
      })}>
      {children}
    </aside>
  );
};
