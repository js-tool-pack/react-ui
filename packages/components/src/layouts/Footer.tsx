import React from 'react';
import { BaseLayoutsProps } from './layout.types';
import { getComponentClass } from '@pkg/shared';
import { getClassNames } from '@tool-pack/basic';

const rootClass = getComponentClass('footer');

export const Footer: React.FC<BaseLayoutsProps> = (props) => {
  const { children, className, ...rest } = props;
  return (
    <footer
      {...rest}
      className={getClassNames(rootClass, {
        [className as string]: className,
      })}>
      {children}
    </footer>
  );
};
