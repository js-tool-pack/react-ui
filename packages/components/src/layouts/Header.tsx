import React from 'react';
import { BaseLayoutsProps } from './layout.types';
import { getComponentClass } from '@pkg/shared';
import { getClassNames } from '@tool-pack/basic';

const rootClass = getComponentClass('header');

export const Header: React.FC<BaseLayoutsProps> = (props) => {
  const { children, className, ...rest } = props;
  return (
    <header
      {...rest}
      className={getClassNames(rootClass, {
        [className as string]: className,
      })}>
      {children}
    </header>
  );
};
