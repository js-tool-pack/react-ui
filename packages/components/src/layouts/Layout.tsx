import React from 'react';
import { getComponentClass } from '@pkg/shared';
import { getClassNames } from '@tool-pack/basic';
import { BaseLayoutsProps } from './layout.types';

const rootClass = getComponentClass('layout');
export const Layout = React.forwardRef<
  HTMLElement,
  BaseLayoutsProps & {
    vertical?: boolean;
  }
>((props, ref) => {
  const { vertical, className, children, ...rest } = props;
  return (
    <section
      {...rest}
      ref={ref}
      className={getClassNames(rootClass, {
        vertical,
        [className as string]: className,
      })}>
      {children}
    </section>
  );
});
