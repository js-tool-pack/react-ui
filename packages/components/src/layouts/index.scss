@import '../scss.variable';

.#{$prefix}layout {
  display: flex;
  flex: 1;
  overflow: hidden;
  &.vertical {
    flex-direction: column;
  }
}
.#{$prefix}aside,
.#{$prefix}main {
  overflow: auto;
}
.#{$prefix}main {
  flex: 1;
}
