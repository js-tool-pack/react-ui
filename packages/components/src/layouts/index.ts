export type { BaseLayoutsProps, LayoutProps } from './layout.types';
export * from './Layout';
