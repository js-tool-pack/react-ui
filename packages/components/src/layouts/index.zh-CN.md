---
category: Components
title: Layout
subtitle: 布局
demo:
  cols: 2
group:
  title: 布局
---

用于页面布局排版。

## 何时使用

项目的 layout 等。

## 代码演示

<!-- prettier-ignore -->
<code src="./demo/all.tsx"></code>

## API

Layout 的属性说明如下：

| 属性     | 说明     | 类型    | 默认值 | 版本 |
| -------- | -------- | ------- | ------ | ---- |
| vertical | 竖向排版 | boolean | false  |      |

支持原生 HTML 的其他所有属性。
