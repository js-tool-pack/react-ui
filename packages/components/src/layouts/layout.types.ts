import React from 'react';

export type BaseLayoutsProps = React.HTMLAttributes<HTMLElement> & {
  tag?: keyof HTMLElementTagNameMap;
};
export type LayoutProps = BaseLayoutsProps & {
  vertical?: boolean;
};
