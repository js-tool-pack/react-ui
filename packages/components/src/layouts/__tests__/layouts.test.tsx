import { testAttrs } from '~/testAttrs';
import { Header, Layout, Main, Footer, Aside } from '~/layouts';
import { render, RenderResult } from '@testing-library/react';

describe('layouts', () => {
  describe('attrs', () => {
    testAttrs(Layout, 'Layout');
    testAttrs(Header, 'Header');
    testAttrs(Main, 'Main');
    testAttrs(Aside, 'Aside');
    testAttrs(Footer, 'Footer');
  });

  test('tag', () => {
    const getChild = (res: RenderResult) => res.container.firstChild;
    expect(getChild(render(<Layout tag="span" />))).toMatchSnapshot();
    expect(getChild(render(<Main tag="span" />))).toMatchSnapshot();
    expect(getChild(render(<Aside tag="span" />))).toMatchSnapshot();
    expect(getChild(render(<Header tag="span" />))).toMatchSnapshot();
    expect(getChild(render(<Footer tag="span" />))).toMatchSnapshot();
  });

  test('vertical', () => {
    const { container } = render(
      <Layout vertical>
        <Header>header</Header>
        <Layout>
          <Aside>aside</Aside>
          <Main>main</Main>
        </Layout>
        <Footer>footer</Footer>
      </Layout>,
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toHaveClass('t-layout--v');
  });
});
