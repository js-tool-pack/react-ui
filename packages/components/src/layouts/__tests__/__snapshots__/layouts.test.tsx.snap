// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layouts tag 1`] = `
<span
  class="t-layout"
/>
`;

exports[`layouts tag 2`] = `
<span
  class="t-main"
/>
`;

exports[`layouts tag 3`] = `
<span
  class="t-aside"
/>
`;

exports[`layouts tag 4`] = `
<span
  class="t-header"
/>
`;

exports[`layouts tag 5`] = `
<span
  class="t-footer"
/>
`;

exports[`layouts vertical 1`] = `
<section
  class="t-layout t-layout--v"
>
  <header
    class="t-header"
  >
    header
  </header>
  <section
    class="t-layout"
  >
    <aside
      class="t-aside"
    >
      aside
    </aside>
    <main
      class="t-main"
    >
      main
    </main>
  </section>
  <footer
    class="t-footer"
  >
    footer
  </footer>
</section>
`;
